docs/source/getting_started.rst

```diff
@@ -103,47 +162,130 @@ Arguments with default values can be annotated as follows:
+The typing module
+*****************
 
-.. note::
+So far, we've added type hints that use only basic concrete types like
+``str`` and ``float``. What if we want to express more complex types,
+such as "a list of strings" or "an iterable of ints"? 
 
-   The earlier stages of mypy, known as the semantic analysis, may
-   report errors even for dynamically typed functions. However, you
-   should not rely on this, as this may change in the future.
+We can find many of these more complex static types inside of the ``typing``
+module. For example, if we want to indicate that some function can accept
+a list of strings, we can use the ``List`` type from the ``typing`` module:
 
-The typing module
-*****************
+.. code-block:: python
+
+   from typing import List
+
+   def greet_all(names: List[str]) -> None:
+       for name in names:
+           print('Hello ' + name)
+
+   names = ["Alice", "Bob", "Charlie"]
+   ages = [10, 20, 30]
+   
+   greet_all(names)   # Ok!
+   greet_all(ages)    # Error due to incompatible types
+
+The ``List`` type is an example of something called a *generic type*: it can
+accept one or more *type parameters*. In this case, we *parameterized* ``List``
+by writing ``List[str]``. This lets mypy know that ``greet_all`` accepts specifically
+lists containing strings, and not lists containing ints or any other type.
+
+In this particular case, our type signature is perhaps a little too rigid.
+After all, there's no reason why this function must accept *specifically* a list --
+it would run just fine if we pass in a tuple, a set, or any custom iterable.
 
-The ``typing`` module contains many definitions that are useful in
-statically typed code. You typically use ``from ... import`` to import
-them (we'll explain ``Iterable`` later in this document):
+We can express this idea by using the ``Iterable`` type instead of ``List``:
 
 .. code-block:: python
 
    from typing import Iterable
 
    def greet_all(names: Iterable[str]) -> None:
        for name in names:
-           print('Hello, {}'.format(name))
+           print('Hello ' + name)
 
-For brevity, we often omit the ``typing`` import in code examples, but
+As a final example, suppose want to write a function that can accept *either*
+ints or strings, but no other types. We can express this using the ``Union`` type:
+
+.. code-block:: python
+
+   from typing import Union
+
+   def normalize_id(user_id: Union[int, str]) -> str:
+       if isinstance(user_id, int):
+           return 'user-{}'.format(100000 + user_id)
+       else:
+           return user_id
+
+The ``typing`` module contains many other useful types. You can find a
+quick overview by looking through the :ref:`mypy cheatsheets <overview-cheat-sheets>` 
+and a more detailed overview (including information on how to make your own
+generic types) by looking through the
+:ref:`type system reference <overview-type-system-reference>`.
+
+One final note: when adding types, the convention is to import types 
```

Comment:
Is it also fair to say that it is conventional to put `typing` imports as the first imports? (It seems like it to me, at least).

If it's not actually a common convention, then it shouldn't derail this PR, but if it is I think it will be nice to mention.

Reply:
I also personally like that style (and I think mypy itself also follows that convention), but I'm thinking it's probably best to not mention this for now -- I'm not sure if other Python linters have caught up on this convention yet. For example, I think isort will sort imports alphabetically?

Reply:
I don't have a strong preference, but I also like `typing` imports first. `isort` however has a different opinion (maybe it is configurable, I have very little experience with it).

Reply:
I don't think that this convention is followed very widely. The `mypy` implementation doesn't consistently put `typing` imports first. I would rather not recommend any particular placement for the imports.